libs/fixture-generators/src/generate-election/generate_election.ts

```diff
@@ -0,0 +1,184 @@
+import {
+  assertDefined,
+  DateWithoutTime,
+  mergeObjects,
+  range,
+} from '@votingworks/basics';
+import {
+  AnyContest,
+  BallotPaperSize,
+  BallotStyle,
+  District,
+  DistrictId,
+  Election,
+  ElectionId,
+  ElectionSchema,
+  Party,
+  PartyId,
+  Precinct,
+  safeParse,
+} from '@votingworks/types';
+import { customAlphabet } from 'nanoid';
+import { defaultConfig, GenerateElectionConfig } from './config';
+import { ballotMeasureText, seal, words } from './source_text';
+
+const generateId = customAlphabet('0123456789abcdefghijklmnopqrstuvwxyz', 12);
+
+function chooseRandom<T>(array: T[]): T {
+  return assertDefined(array[Math.floor(Math.random() * array.length)]);
+}
+
+function chooseRandomSubset<T>(array: T[]): T[] {
+  const numItems = Math.floor(Math.random() * array.length);
+  const result = new Set<T>();
+  while (result.size < numItems) {
+    result.add(chooseRandom(array));
+  }
+  return Array.from(result);
+}
+
+function randomString(maxLength: number, sourceWords: string[]): string {
+  let result = `${maxLength}:`;
+  while (result.length < maxLength) {
+    result += ` ${chooseRandom(sourceWords)}`;
+  }
+  return result.slice(0, maxLength);
+}
+
+type PartialDeep<T> = T extends object
+  ? {
+      [P in keyof T]?: PartialDeep<T[P]>;
+    }
+  : T;
+
+/**
+ * Generates an election with mock content based on the given parameters.
+ */
+export function generateElection(
+  inputConfig: PartialDeep<GenerateElectionConfig>
+): Election {
+  const config = mergeObjects(
+    defaultConfig as Partial<GenerateElectionConfig>,
+    inputConfig
+  ) as GenerateElectionConfig;
+  const { maxStringLengths } = config;
+
+  function generateDistrict(index: number): District {
+    return {
+      id: `district-${index}` as DistrictId,
+      name: randomString(maxStringLengths.districtName, words.locations),
+    };
+  }
+  const districts = range(0, config.numDistricts).map(generateDistrict);
+
+  function generatePrecinct(index: number): Precinct {
+    return {
+      id: `precinct-${index}`,
+      name: randomString(maxStringLengths.precinctName, words.locations),
+    };
+  }
+  const precincts = range(0, config.numPrecincts).map(generatePrecinct);
+
+  function generateBallotStyle(index: number): BallotStyle {
```

Comment:
I wonder if this will create wacky elections with failures we wouldn't see in the real world by having districts/precincts without corresponding ballot styles

Reply:
Good point. I'd be down to see if that happens just out of curiosity. But I do think I added a filter in VxDesign to prevent that, so maybe it's not so relevant to test.